app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/CloudLoaderAsyncTask.java

```diff
@@ -0,0 +1,289 @@
+package com.amaze.filemanager.asynchronous.asynctasks;
+
+import android.database.Cursor;
+import android.os.AsyncTask;
+import android.support.annotation.NonNull;
+import android.widget.Toast;
+
+import com.amaze.filemanager.R;
+import com.amaze.filemanager.activities.MainActivity;
+import com.amaze.filemanager.database.CloudHandler;
+import com.amaze.filemanager.database.models.CloudEntry;
+import com.amaze.filemanager.exceptions.CloudPluginException;
+import com.amaze.filemanager.utils.DataUtils;
+import com.amaze.filemanager.utils.OpenMode;
+import com.amaze.filemanager.utils.application.AppConfig;
+import com.cloudrail.si.CloudRail;
+import com.cloudrail.si.exceptions.AuthenticationException;
+import com.cloudrail.si.exceptions.ParseException;
+import com.cloudrail.si.interfaces.CloudStorage;
+import com.cloudrail.si.services.Box;
+import com.cloudrail.si.services.Dropbox;
+import com.cloudrail.si.services.GoogleDrive;
+import com.cloudrail.si.services.OneDrive;
+
+import java.lang.ref.WeakReference;
+
+public class CloudLoaderAsyncTask extends AsyncTask<Void, Void, Boolean> {
+
+    private Cursor data;
+    private WeakReference<MainActivity> mainActivityWeakReference;
+    private CloudHandler cloudHandler;
+    private DataUtils dataUtils;
+    public CloudLoaderAsyncTask(WeakReference<MainActivity> mainActivityWeakReference,
+                                CloudHandler cloudHandler, Cursor data) {
+        this.data = data;
+        this.mainActivityWeakReference = mainActivityWeakReference;
+        this.cloudHandler = cloudHandler;
+        this.dataUtils = DataUtils.getInstance();
+    }
+
+    @Override
+    @NonNull
+    public Boolean doInBackground(Void... voids) {
+        boolean hasUpdatedDrawer = false;
+
+        if (data == null) return false;
+        if (data.getCount() > 0 && data.moveToFirst()) {
+            do {
+                if (mainActivityWeakReference.get() == null) cancel(true);
+                if (isCancelled()) cancel(true);
```

Comment:
> Attempts to cancel execution of this task. This attempt will fail if the task has already completed, already been cancelled, or could not be cancelled for some other reason.
> [...]

-- [AsyncTask.cancel() documentation](https://developer.android.com/reference/android/os/AsyncTask.html#cancel(boolean))

And you still need to exit the function if you want to.